src/pudl/extract/epacems.py

```diff
@@ -146,29 +149,73 @@ def _csv_to_dataframe(self, csv_file) -> pd.DataFrame:
+    # Could maybe replace our pydantic settings models with
+    # Dagster Types that check working partitions and tables.
+    # https://docs.dagster.io/concepts/types#defining-a-dagster-type
+    # Might be cumbersome though.
+    years = context.op_config["years"]
+    states = context.op_config["states"]
+    settings = EpaCemsSettings(years=years, states=states)
+
+    # Verify that we have a PUDL DB with plant attributes:
+    pudl_engine = context.resources.pudl_engine
+    inspector = sa.inspect(pudl_engine)
+    if "plants_eia860" not in inspector.get_table_names():
+        raise RuntimeError(
+            "No plants_eia860 available in the PUDL DB! Have you run the ETL? "
+            f"Trying to access PUDL DB: {pudl_engine}"
+        )
+
+    eia_plant_years = pd.read_sql(
+        """
+        SELECT DISTINCT strftime('%Y', report_date)
+        AS year
+        FROM plants_eia860
+        ORDER BY year ASC
+        """, pudl_engine).year.astype(int)
+
+    missing_years = list(set(settings.years) - set(eia_plant_years))
+    if missing_years:
+        logger.info(
+            f"EPA CEMS years with no EIA plant data: {missing_years} "
+            "Some timezones may be estimated based on plant state."
+        )
+
+    # Yield the partitions once everything is validated.
+    for year in settings.years:
+        for state in settings.states:
+            yield DynamicOutput(EpaCemsPartition(state=state, year=year), mapping_key=f"{state}{year}")
+
+
+@op(required_resource_keys={"datastore"})
+def extract(context, partition: EpaCemsPartition) -> pd.DataFrame:
     """
     Coordinate the extraction of EPA CEMS hourly DataFrames.
 
     Args:
-        epacems_years (list): The years of CEMS data to extract, as 4-digit
-            integers.
-        states (list): The states whose CEMS data we want to extract, indicated
-            by 2-letter US state codes.
-        ds (:class:`Datastore`): Initialized datastore
+        partition: A single EPA CEMS partitions (year, state).
 
     Yields:
         pandas.DataFrame: A single state-year of EPA CEMS hourly emissions data.
 
     """
-    ds = EpaCemsDatastore(ds)
-    for year in epacems_years:
-        for state in states:
-            partition = EpaCemsPartition(state=state, year=year)
-            logger.info(f"Processing EPA CEMS hourly data for {state}-{year}")
-            # We have to assign the reporting year for partitioning purposes
-            df = (
-                ds.get_data_frame(partition)
-                .assign(year=year)
-            )
-            yield df
+    ds = EpaCemsDatastore(context.resources.datastore)
+    logger.info(
+        f"Processing EPA CEMS hourly data for {partition.state}-{partition.year}")
+    # We have to assign the reporting year for partitioning purposes
+    df = (
+        ds.get_data_frame(partition)
+        .assign(year=partition.year)
+    )
+    return Output(df, metadata={"Number of Rows": len(df)})
```

Comment:
Why can/should this be `return` instead of `yield`?

Reply:
I think because `extract()` is handling one partition at a time where as `gather_partitions()` is yielding multiple partitions. I was roughly following their Dynamic Mapping [example](https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs#dynamic-mapping--collect).